src/iceberg/util/type_util.h

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+#include <functional>
+#include <stack>
+
+#include "iceberg/type.h"
+
+/// \file iceberg/util/type_util.h
+/// Utility functions for Iceberg types.
+
+namespace iceberg {
+
+/// \brief Index parent field IDs for all fields in a struct hierarchy.
+/// \param root_struct The root struct type to analyze
+/// \return A map from field ID to its parent struct field ID
+std::unordered_map<int32_t, int32_t> indexParents(const StructType& root_struct) {
+  std::unordered_map<int32_t, int32_t> id_to_parent;
+  std::stack<int32_t> parent_id_stack;
+
+  // Recursive function to visit and build parent relationships
+  std::function<void(const Type&)> visit;
+  visit = [&](const Type& type) {
+    switch (type.type_id()) {
+      case TypeId::kStruct: {
+        const auto& struct_type = static_cast<const StructType&>(type);
+        for (const auto& field : struct_type.fields()) {
+          if (!parent_id_stack.empty()) {
+            id_to_parent[field.field_id()] = parent_id_stack.top();
```

Comment:
`field.field_id()` might be invalid or duplicates with other fields if we don't validate the input `root_struct`. If you don't want to change the return type to be `Result<...>`, perhaps add a comment to let readers know what is the expected input. Otherwise we need to validate `field.field_id()` is non-negative and `id_to_parent.emplace(field.field_id(), parent_id_stack.top())` is successful (i.e. not exists already).